addon/webextension/background/analytics.js

```diff
@@ -77,5 +110,111 @@ this.analytics = (function() {
+  let rules = [{
+    name: 'page-action',
+    start: { action: 'start-shot', label: 'toolbar-button' },
+    end: { action: 'internal', label: 'unhide-preselection-frame' },
+    cancel: [{ action: 'cancel-shot' }]
+  }, {
+    name: 'context-menu',
+    start: { action: 'start-shot', label: 'context-menu' },
+    end: { action: 'internal', label: 'unhide-preselection-frame' },
+    cancel: [{ action: 'cancel-shot' }]
+  }, {
+    name: 'capture-full-page',
+    start: { action: 'capture-full-page' },
+    end: { action: 'internal', label: 'unhide-preview-frame' },
+    cancel: [{ action: 'cancel-shot' }]
+  }, {
+    name: 'capture-visible',
+    start: { action: 'capture-visible' },
+    end: { action: 'internal', label: 'unhide-preview-frame' },
+    cancel: [{ action: 'cancel-shot' }]
+  }, {
+    name: 'make-selection',
+    start: { action: 'make-selection' },
+    end: { action: 'internal', label: 'unhide-selection-frame' },
+    cancel: [{ action: 'cancel-shot' }]
+  }, {
+    name: 'save-shot',
+    start: { action: 'save-shot' },
+    end: { action: 'internal', label: 'open-shot-tab' },
+    cancel: [{ action: 'cancel-shot' }, { action: 'upload-failed' }]
+  }, {
+    name: 'save-visible',
+    start: { action: 'save-visible' },
+    end: { action: 'internal', label: 'open-shot-tab' },
+    cancel: [{ action: 'cancel-shot' }, { action: 'upload-failed' }]
+  }, {
+    name: 'save-full-page',
+    start: { action: 'save-full-page' },
+    end: { action: 'internal', label: 'open-shot-tab' },
+    cancel: [{ action: 'cancel-shot' }, { action: 'upload-failed' }]
+  }, {
+    name: 'save-full-page-truncated',
+    start: { action: 'save-full-page-truncated' },
+    end: { action: 'internal', label: 'open-shot-tab' },
+    cancel: [{ action: 'cancel-shot' }, { action: 'upload-failed' }]
+  }, {
+    name: 'download-shot',
+    start: { action: 'download-shot' },
+    end: { action: 'internal', label: 'deactivate' },
+    cancel: [{ action: 'cancel-shot' }]
+  }, {
+    name: 'download-full-page',
+    start: { action: 'download-full-page' },
+    end: { action: 'internal', label: 'deactivate' },
+    cancel: [{ action: 'cancel-shot' }]
+  }, {
+    name: 'download-full-page-truncated',
+    start: { action: 'download-full-page-truncated' },
+    end: { action: 'internal', label: 'deactivate' },
+    cancel: [{ action: 'cancel-shot' }]
+  }, {
+    name: 'download-visible',
+    start: { action: 'download-visible' },
+    end: { action: 'internal', label: 'deactivate' },
+    cancel: [{ action: 'cancel-shot' }]
+  }];
+
+  // Match a filter (action and optional label) against an action and label.
+  function match(filter, action, label) {
+    return filter.label ?
+      filter.action === action && filter.label === label :
+      filter.action === action;
+  }
+
+  function anyMatches(filters, action, label) {
+    let matched = false;
+    filters.forEach(filter => {
+      matched = matched || match(filter, action, label);
+    });
```

Comment:
Nitpick: since we are only interested in the first matching filter, we can use `array.find()` here.